### ISSUE TITLE
“show_batch_normalization_layer” assert layer_axis error

### ISSUE THREAD

**Issue by taotaoyuhust** [Dobiasd/frugally-deep#109]
during load model 
report an error in func  show_batch_normalization_layer

 assert layer_axis == -1 or layer_axis + 1 == len(layer.input_shape)
where  layer_axis is 1, and layer.input_shape is (242,242,32)
what should i do? 3x!

**Comment by Dobiasd** [comment]
Are you using the latest version of Keras and of frugally-deep?

Can you provide a minimal example to reproduce your problem?

**Comment by taotaoyuhust** [comment]
my keras version is 2.2.2  
i am using
self.model.load_weights(self.config.hdf5_path + 'simple_epoch08.h5')
self.model.save(self.config.hdf5_path + 'keras_model.h5', include_optimizer=False) 
to produce the keras_model.h5

 my model is this :
[models.zip](https://github.com/Dobiasd/frugally-deep/files/2702312/models.zip)

thanks a lot!

**Comment by Dobiasd** [comment]
The issue is, the `BatchNormalization` layer in your model uses `axis == 1`, which is not supported with a four-dimensional `layer.input_shape`, as in that case.

But even if we were able to mitigate this issue, your model still would not work, because it uses `Conv2DTranspose` layers, which are not supported by frugally-deep.

In case you still want/need to use frugally-deep nonetheless, you might try to change your model's architecture to using `UpSampling2D + Conv2D` instead of `Conv2DTranspose`, which according to [this article](https://distill.pub/2016/deconv-checkerboard/), seems to be the better approach anyways.

**Comment by taotaoyuhust** [comment]
i still need to use frugally-deep,  but i actually don't know how to replace the Conv2DTranspose layer and also means i have to retrain the whole net....

**Comment by Dobiasd** [comment]
You can not replace `Conv2DTranspose` with `UpSampling2D` and `Conv2D` in an already trained net. This only works if you have control over the model's architecture and also train it on your own. In case you are using a pre-trained model, you are of course bound to use what is in it.

Assuming this is the case, i.e., you can't get rid of `Conv2DTranspose` easily, frugally-deep is currently not the right tool for you. Maybe you can convert your Keras model to a raw TensorFlow model and then use [TensorFlow Lite](https://www.tensorflow.org/lite/) to deploy it?

> i still need to use frugally-deep

What brings you to this conclusion?

**Comment by Dobiasd** [comment]
Did you use "release" mode (in contrast to "debug" mode) in Visual C++? Accidentally using "debug" mode is a common cause for performance problems there.

**Comment by taotaoyuhust** [comment]
Thank u， I will try it！

**Comment by Dobiasd** [comment]
Did it help?

**Comment by taotaoyuhust** [comment]
thanks  for your attention, i finally  resolved this problem using another version of tensorflow.dll ,now it works for me , i can using the model of pb format in vs2015!
btw ,i will be very appreciated to see you add the Conv2DTranspose  layer：）

**Comment by Dobiasd** [comment]
Thanks for the response. I'll thus close the issue.

---

> btw ,i will be very appreciated to see you add the Conv2DTranspose layer：）

Up to now I still speculate on `Conv2DTranspose` simply [becoming to be out of fashion](https://distill.pub/2016/deconv-checkerboard/). :wink:
